charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+
+import pytest
+import tenacity
+import yaml
+from httpx import HTTPStatusError
+from lightkube import ApiError, Client, codecs
+from lightkube.generic_resource import (
+    create_namespaced_resource,
+    load_in_cluster_generic_resources,
+)
+from lightkube.resources.apiextensions_v1 import CustomResourceDefinition
+from lightkube.resources.core_v1 import PersistentVolumeClaim, Pod, Service
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+ASSETS_DIR = Path("tests") / "assets"
+PVC_TEMPLATE_FILE = ASSETS_DIR / "dummy-pvc.yaml.j2"
+TENSORBOARD_TEMPLATE_FILE = ASSETS_DIR / "dummy-tensorboard.yaml.j2"
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+PVC_NAME = "dummy-pvc"
+TENSORBOARD_NAME = "dummy-tensorboard"
+TENSORBOARD_RESOURCE = create_namespaced_resource(
+    group="tensorboard.kubeflow.org",
+    version="v1alpha1",
+    kind="tensorboard",
+    plural="tensorboards",
+)
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build the charm-under-test and deploy it together with related charms.
+
+    Assert on the unit status before any relations/configurations take place.
+    """
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+    tb_controller_image = METADATA["resources"]["tensorboard-controller-image"]["upstream-source"]
+    resources = {"tensorboard-controller-image": tb_controller_image}
+
+    await ops_test.model.deploy(charm, resources=resources, application_name=APP_NAME, trust=True)
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="waiting", raise_on_blocked=True, timeout=1000
```

Comment:
Should this ever really take 15 minutes?  If yes, leave it as is.  But if not, let's reduce so we don't wait on our CI for so long.  My guess is 5 minutes is plenty

Reply:
This is a good point! We use these huge, arbitrary timeouts in our charm tests a lot, and I was actually wondering whether there's some rationale behind setting them. I do agree that this should never take more than 5', but let's keep a note and discuss about establishing some rules for uniformity across our apps.